presto-array/src/main/java/com/facebook/presto/array/ReferenceCountMap.java

```diff
@@ -59,19 +64,44 @@ public long sizeOf()
         return INSTANCE_SIZE + SizeOf.sizeOf(key) + SizeOf.sizeOf(value) + SizeOf.sizeOf(used);
     }
 
-    private static final class ObjectStrategy
-            implements Strategy<Object>
+    /**
+     * Get the 64-bit hash code for an object
+     */
+    private static long getHashCode(Object key)
     {
-        @Override
-        public int hashCode(Object object)
-        {
-            return System.identityHashCode(object);
+        // identityHashCode does not guarantee the uniqueness of an object. We need some extra identity.
+        // In the most cases below, we use size of an object to be the extra identity.
+        // Experiments show that with 100 million objects created, using identityHashCode has a collision rate around 2.5%.
+        // However, if these 100 million objects are combined with 10 different sizes, the collision rate is around 0.1%.
+        // The collision rate can be lower than 0.001% if there are 1000 different sizes.
+        int extraIdentity;
+        if (key instanceof Block) {
+            extraIdentity = (int) ((Block) key).getRetainedSizeInBytes();
         }
-
-        @Override
-        public boolean equals(Object left, Object right)
-        {
-            return left == right;
+        else if (key instanceof Slice) {
+            extraIdentity = System.identityHashCode(((Slice) key).getBase());
```

Comment:
base can be `null`, also a slice of a slice will share the same base object.  Is that what you want?  If not, I'd use size.